sdk/opa.go

```diff
@@ -0,0 +1,426 @@
+// Copyright 2021 The OPA Authors.  All rights reserved.
+// Use of this source code is governed by an Apache2
+// license that can be found in the LICENSE file.
+
+// Package sdk contains a high-level API for embedding OPA inside of Go programs.
+package sdk
+
+import (
+	"bytes"
+	"context"
+	"crypto/rand"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/open-policy-agent/opa/ast"
+	"github.com/open-policy-agent/opa/bundle"
+	"github.com/open-policy-agent/opa/internal/ref"
+	"github.com/open-policy-agent/opa/internal/uuid"
+	"github.com/open-policy-agent/opa/logging"
+	"github.com/open-policy-agent/opa/metrics"
+	"github.com/open-policy-agent/opa/plugins"
+	"github.com/open-policy-agent/opa/plugins/discovery"
+	"github.com/open-policy-agent/opa/plugins/logs"
+	"github.com/open-policy-agent/opa/rego"
+	"github.com/open-policy-agent/opa/server"
+	"github.com/open-policy-agent/opa/storage"
+	"github.com/open-policy-agent/opa/storage/inmem"
+	"github.com/open-policy-agent/opa/topdown/cache"
+)
+
+// OPA represents an instance of the policy engine. OPA can be started with
+// several options that control configuration, logging, and lifecycle.
+type OPA struct {
+	id      string
+	state   *state
+	mtx     sync.Mutex
+	logger  logging.Logger
+	console logging.Logger
+	config  []byte
+}
+
+type state struct {
+	manager                *plugins.Manager
+	interQueryBuiltinCache cache.InterQueryCache
+	queryCache             *queryCache
+}
+
+// New returns a new OPA object. This function should minimally be called with
+// options that specify an OPA configuration file.
+func New(ctx context.Context, opts Options) (*OPA, error) {
+
+	id, err := uuid.New(rand.Reader)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := opts.init(); err != nil {
+		return nil, err
+	}
+
+	opa := &OPA{
+		id: id,
+		state: &state{
+			queryCache: newQueryCache(),
+		},
+	}
+
+	opa.config = opts.config
+	opa.logger = opts.Logger
+	opa.console = opts.ConsoleLogger
+
+	return opa, opa.configure(ctx, opa.config, opts.Ready, opts.block)
+}
+
+// Configure updates the configuration of the OPA in-place. This function should
+// be called in response to configuration updates in the environment. This
+// function is atomic. If the configuration update cannot be successfully
+// applied, the old configuration will remain intact.
+func (opa *OPA) Configure(ctx context.Context, opts ConfigOptions) error {
+
+	if err := opts.init(); err != nil {
+		return err
+	}
+
+	// NOTE(tsandall): In future we could be more intelligent about
+	// re-configuration and avoid expensive background processing.
+	opa.mtx.Lock()
+	equal := bytes.Equal(opts.config, opa.config)
+	opa.mtx.Unlock()
+
+	if equal {
+		close(opts.Ready)
+		return nil
+	}
+
+	return opa.configure(ctx, opts.config, opts.Ready, opts.block)
+}
+
+func (opa *OPA) configure(ctx context.Context, bs []byte, ready chan struct{}, block bool) error {
```

Comment:
[nit] I guess if it's only for closing, `<-chan struct{}` would do, too?

Reply:
"cannot close a receive-only channel" -- not sure what I was thinking last night :D